### ISSUE TITLE
A few thoughts on IPFS Cluster

### ISSUE THREAD

**Issue by sashahilton00** [ipfs/ipfs-cluster#584]
So I've been reading up on IPFS-Cluster for the better part of 2 hours now, and I think I understand it's function as a replication mechanism for IPFS. I have the following questions and thoughts:

- Isn't IPFS-Cluster essentially doing the same thing as filecoin will be doing, minus the incentivisation from earning currency? This isn't a criticism; I can fully see the role of this in closed/private networks, but a clarification would be great.
- Assuming the point above about it's usage as a coordinated private network, what are the plans for access control? It is my understanding that any peer in the network has full read/write over the content pinned to other peers.

If this is the case, might I propose that it could be highly beneficial to integrate the authentication/acl mechanisms of ipfs-cluster with ssl certificates? From where I'm sitting, SSL provides a great many features: with a root CA signing client certificates, clients authenticate to the cluster by presenting their client certificate, which is verified against the network CA. read/write priviledges could similarly be restricted through the use of certificate entitlements, allowing one to permission commands such as pin to cluster, delete file, etc. Furthermore, there are a number of other things like CRL which could be used to retrospectively revoke access without needing to change and redistribute the secret key.

Open to suggestions/thoughts...

**Comment by NatoBoram** [comment]
One thing I want from IPFS-Cluster is being able to publish a key that volunteers can enter to be part of my cluster and donate a few gigs of storage to my cluster without gaining write access. This would allow for a community-run hosting service.

**Comment by sashahilton00** [comment]
that's one of the main features I'm looking for as well. what ideally I'd have is a cluster whereby all peers can upload files, but only whitelisted peers were able to issue cluster commands.

**Comment by sashahilton00** [comment]
Excerpts from an IRC conversation:

```

[02:25:23] lanzafame:	fyi I have started on a response to your issue, just haven't had a chance to finish it
[02:31:37] sashahilton:	the cheating aspect is interesting, hadn't considered that, a bit like the bittorrent fake-block attack. I assume the solution to that would be similar to their solution to the attack; blacklist bad peers
[02:34:22] lanzafame:	filecoin's solution is the proofs of replication and spacetime, clusters solution is only peers with the cluster secret are allowed to pin, so only give the secret to peers you trust
[02:34:52] sashahilton:	with regards to authorization, that does sound interesting, though obviously comes after getting consensus stable. I haven't had the chance yet, but i plan on testing cluster with high latency links, as the ultimate use case is permissioned file sharing across peers distributed globally. are there any algorithms currently being proposed? i read somewhere that cluster uses raft at the moment, which as I understand it is
[02:34:52] sashahilton:	 meant for low latency performance
[02:37:15] sashahilton:	This should be flexible enough to support the authorization scenario that you mention. i.e. everyone can pin, only trusted peers can unpin, rpc requests to other peers are allowed to read-only endpoints etc. that's similar to the use case that i'm looking for, with the small difference that I'd like to run a private network, which allows anyone to add/locally pin files, but only peers on the whitelist to be able to
[02:37:15] sashahilton:	 control cluster behaviour, including cluster pinning and unpinning.
[02:37:17] lanzafame:	hsanjuan is leading the charge on using CRDTs as the 'consensus' algorithm, how that is going to look I'm not sure at the moment
[02:38:12] sashahilton:	@lanzafame is there a spec that is wip that I can read up on the proposed usage of CRDTs on, or is that informal planning atm?
[02:38:21] postables:	sashahilton: you can "achieve" that in some sense by adding a permissioned API layer ontop of your cluster
[02:42:18] postables:	firewall -> API Service (performs permission checks, requires user accounts, etc...) -> Connect to IPFS Cluster <- is how i do it right now. The API service will connect to my IPFS Cluster API
[02:48:55] sashahilton:	ok. I can see how that would work, but it doesn't solve the problem behind the use case I have, which in simple terms is that I have, say, 80TB of data, and wish to distribute an application that bundles ipfs-cluster, such that all users of the application act as untrusted peers, each hosting 10GB of data or whatever
[02:54:07] sashahilton:	so in the current state, one needs to wait for a new consensus algorithm to be implemented that isn't as strict on peersets and latency, likely via CRDTs. Once that's accomplished, authorization gets implemented, allowing whitelisted peers to manage the cluster via signed commands. That leaves one other small question. Would access revocation be part of that? i.e. would one be able to remove trusted peers from the
[02:54:07] sashahilton:	 whitelist through the ipfs-cluster, or would that be managed externally?
[02:56:26] lanzafame:	So you are after a managed federation mechanism, correct?
[02:57:31] sashahilton:	essentially, yes. I'm not looking for a full blown distributed federation server, but a means to grant and revoke access to cluster management endpoints by priviledged peers through the cluster would be awesome.
[03:01:09] postables:	but yea right now I don't think there's a practical way to integrate untrusted peers with ipfs cluster
[03:01:44] postables:	IMO if you want to achieve functionality similar to that (minus untrusted peers) you're best with an RBAC style permission layer to control what functions people can/can't execute on your API server
[03:01:44] @lanzafame:	Yeah the volunteer data replication is a use case that we are working hard to support. There is just some difficulty in how to deal with if a volunteer disappearing with the 10% of a dataset that they were replicating
```

**Comment by NatoBoram** [comment]
> That potentially lets you run your idea with a bunch of trusted volunteers.

I don't want to trust volunteers, and I'm pretty sure any system that needs to trust user inputs is going to eat dirt.

Let's imagine a real use case with `collaborative pinsets`. I talked to r/LineageOS about their storage problem and how to solve it. Here's how it currently is.

1. They have a limit of 1 TB to store LineageOS builds
2. After 4 weeks, builds are removed
3. Devices that didn't get updates in 4 weeks are dropped from support

The thing is their builds are still good even if they don't have the latest shiniest security fix from Google simply because they have *more* than what the stock firmware will offer. In short, they need more space to store, at least, the latest build from every single device they've ever supported.

It would be awesome to be able to donate disk space like that. If they stored the builds in an IPFS Cluster of 1 TB and published a "volunteer's key", and people could just hop in to be part of the cluster without gaining write access, that would mean they're able to overcome their storage shortage.

That's something that could be achieved with FileCoin, but they'd have to pay for the storage, it wouldn't be a community effort.

---

Similarly, I want to keep track of a pinset, but all my computers are getting reinstalled pretty often. If I could create an IPFS Cluster from a machine that's essentially just an orchestrator, and use volunteer keys with all my other machines, I would create a pretty impressive cluster and seed files while promoting IPFS in forums and everywhere I share files. Right now, you have to unlink a disconnected machine before you can add a new one, so it's pretty frustrating and the effort is actually more than what's worth.

---

In short, I'm really excited by collaborative pinsets and I have both public and private projects with it as soon as it's available.

And, to be fair, I don't even need consistency of the content, just the hashes. I'm okay with some alpha version.

**Comment by sashahilton00** [comment]
Agreed, keeping track of a pinset would definitely be desirable, as in an archival scenario with volunteers, it's going to be the case that servers are joining and leaving relatively frequently.

The reason I mentioned OpenSSL certs previously is because it already contains a mechanism for whitelisting and revoking access on a per client basis. From what I've seen thus far, IPFS has little by way of ACL, thus my thinking was that by (optionally) adding OpenSSL for cluster ACL, one can have a CA that signs client certificates for authentication, and maintains a CRL endpoint for revocation. One could alternatively implement ACL using the current ipfs tools by encrypting files before uploading to the ipfs network, then having an api server which serves decryption keys to authenticated users on top of that.

I guess it boils down to what ipfs-cluster is looking to become. If it is looking to compete with other file sharing protocols like FTP/SMB/etc. Then the current ACL probably isn't sufficient, but if on the other hand it's looking to replace torrents and other forms of decentralised storage only, then ACL probably isn't a consideration.

A potential halfway house might be to create a spec for an auth server that authenticates peers and serves (per file) decryption keys/uploads encryption keys that is built in to ipfs cluster. That way, one could pass in an option like `--federation-server authserver.com` or similar, and have ipfs-cluster download the file from the cluster, decrypt the chunks from the network to retrieve the encrypted file, then pull the decryption key from the federation server to decrypt the file to it's original form, all in a transparent manner which negates the requirement for build your own encryption on top of ipfs-cluster if you need ACL.

**Comment by sashahilton00** [comment]
I think the real benefit that IPFS Cluster could bring is to deposit chunks across the network of peers as opposed to each peer hosting an entire copy of a site. Take any kind of music/video site, for example. If each peer has to maintain a full copy of each file, then it quickly becomes prohibitive for most peers to do so.

Not sure how familiar you guys are with Storj, but the ideal system for the archival use case seems to be a cluster comprised of both trusted and volunteer nodes, with the pinned content spread evenly across them, each chunk replicated 2-3 times to ensure the ability to auto repair the network if a node (and a chunk of content) go offline. Incentivisation through the likes of Filecoin isn't really all that necessary, but it would be good to have a node reputation system, whereby if a node remains online for a consecutive number of hours, has low latency, etc. they get more chunks to store than a node which has only been online for a few hours, or has high latency/packet loss.

The question is, how decentralised does decentralised need to be? I'd personally go for a system whereby each node acted as a server in the decentralised network, and subscribed to an unlimited number of federation servers. the federation servers would then manage chunk distribution in that federation's cluster, monitor metrics sent by the nodes, etc. Individuals could then join or leave federation servers as they saw fit, or allocate X much space to server X, etc. meaning that people could actively choose to host content they support as opposed to every site they visit. At this point i'm just thinking aloud. I think this is far beyond the scope of what IPFS cluster is trying to do. Let's just get collaborative peersets working first :)

**Comment by hsanjuan** [comment]
I will close this for now (doing issue clean-up and this is not actionable). I'd encourage any further discussion to happen on https://discuss.ipfs.io . Cheers!
